### PR TITLE
[2/N] Added KDLoss based AutoQuantize

### PR DESCRIPTION
**Pull request by realAsma**
## What does this PR do?

**Type of change:** ? New Feature

**Overview:** 

This PR extends AutoQuantize with KL Divergence Loss-based sensitivity measurement as an alternative to the existing gradient-based approach. KD Loss uses a binary searcher similar to the binary searcher in FastNAS.

AutoQuantize gradient is faster than KL Divergence based AutoQuantize. However KL Divergence does not need the model implementation to support gradient backward. In addition, AutoQuantize collected KL Divergence is useful for sensitivity analysis of the model. KL Divergence is a more direct measure of sensitivity than gradient scores.

## Usage

see `tests/unit/torch/quantization/test_autoquant.py`

## Testing

Testes with unit tests.

Result for Qwen3 8B
image

## Before your PR is "*Ready for review*"

- **Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/CONTRIBUTING.md)** and your commits are signed.
- **Is this change backward compatible?**: Yes/No 
- **Did you write any new necessary tests?**: Yes/No
- **Did you add or update any necessary documentation?**: Yes/No
- **Did you update [Changelog](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/CHANGELOG.rst)?**: Yes/No

### CODE REVIEW

File: modelopt/torch/quantization/algorithms.py [hunk 1]
```diff
@@ -170,19 +171,35 @@ class QuantRecipeHparam(Hparam):
...
+    * Keeps track of the importance of each recipe in a dict instead of a tensor.
+    * Registers itself with each score_module via the _hparams_for_scoring attribute.
     """
 
     def __init__(
         self,
         choices: Sequence[QuantRecipe] | None = None,
-        nn_modules: list[nn.Module] | None = None,
+        quant_modules: list[nn.Module] | None = None,
+        score_modules: list[nn.Module] | None = None,
```
Comment (meenchen) [inline modelopt/torch/quantization/algorithms.py:185]: qq, what are the use cases for specifying score_modules?

Comment (realAsma) [reply modelopt/torch/quantization/algorithms.py:185]: Score_module is used only for gradient AutoQuantize. 

By default, for each quant module, its gradient based sensitivity score is estimated using that module's output perturbation.
However, the sensitivity can also be estimated by looking at perturbation at a separate point in the neural
network (score module). This is helpful in some cases such as MoEs for speed and lower memory consumption.
Since all experts are already restricted to the same quant format by quant grouping rules, their sensitivity
can be estimated together at a single point (e.g., the MLP output level).

see - https://github.com/NVIDIA/TensorRT-Model-Optimizer/pull/586/files#r2551002333

File: modelopt/torch/quantization/algorithms.py [hunk 2]
```diff
@@ -963,8 +963,231 @@ def run_search_with_stats(self, max_weight_size, verbose=False):
...
+            logits, parallel_state.tensor_parallel_group.group, return_log_prob
+        )
+    return _get_softmax(logits, return_log_prob)
+
+
+def _get_kl_div_loss(
+    prob_unquant: torch.Tensor, logits_quant: torch.Tensor, lm_head: nn.Module = None
+) -> torch.Tensor:
+    log_prob_quant = _get_prob_from_logits(logits_quant, return_log_prob=True, lm_head=lm_head)
+    # We dont need to calculate the full kl div loss here, just get p*log_q
```
Comment (meenchen) [inline modelopt/torch/quantization/algorithms.py]: Is it because we are takeing p * log(p) as contant for each recipe?

File: modelopt/torch/quantization/algorithms.py [hunk 3]
```diff
@@ -963,8 +963,231 @@ def run_search_with_stats(self, max_weight_size, verbose=False):
...
+                list(named_hparams(self.model, configurable=True)), desc="Evaluating hparams"
+            ):
+                if not isinstance(hparam, QuantRecipeHparam):
+                    continue
+                for recipe in hparam.choices:
+                    if recipe == QuantRecipe(quant_cfg=None):
+                        continue
+                    hparam.active = recipe
+                    logits_quant = self.config["forward_step"](self.model, data)
+                    score = _get_kl_div_loss(prob_unquant, logits_quant, _get_lm_head(self.model))
```
Comment (meenchen) [inline modelopt/torch/quantization/algorithms.py:1159]: So the loss here is sum(p * log_q). Do we need to flip the sign bit? I think the original formula is sum(p * log_p) - sum(p * log_q).

File: modelopt/torch/quantization/algorithms.py [hunk 4]
```diff
@@ -963,8 +963,231 @@ def run_search_with_stats(self, max_weight_size, verbose=False):
...
+
+        # Initialize binary search bounds
+        min_score = min(all_scores)
+        max_score = max(all_scores)
+        threshold = (min_score + max_score) / 2.0
+        lower_bound = min_score
+        upper_bound = max_score
+
+        # Run for fixed number of iterations
+        max_iterations = 100
```
Comment (meenchen) [inline modelopt/torch/quantization/algorithms.py:1189]: Do we need to make this as a parameter?

File: modelopt/torch/quantization/algorithms.py [hunk 5]
```diff
@@ -963,8 +963,231 @@ def run_search_with_stats(self, max_weight_size, verbose=False):
...
+            for name in self.candidate_stats:
+                formats = self.candidate_stats[name]["formats"]
+                scores = self.candidate_stats[name]["scores"]
+                costs = self.candidate_stats[name]["costs"]
+
+                selected_idx = 0
+                for idx in range(len(formats)):
+                    if scores[idx] <= threshold:
+                        selected_idx = idx
+                        break
```
Comment (meenchen) [inline modelopt/torch/quantization/algorithms.py:1210]: Will we always find a format that satisfies the threshold?

File: modelopt/torch/quantization/algorithms.py (hunk 4 above)
Comment (realAsma) [reply modelopt/torch/quantization/algorithms.py:1189]: For now I suggest not since the search does not change much after first 10-20 iterations (binary search). I just put a large enough number - 100 CPU for loop iterations are not super cheap.

File: modelopt/torch/quantization/algorithms.py (hunk 5 above)
Comment (realAsma) [reply modelopt/torch/quantization/algorithms.py:1210]: yes, We always have None calibration which has a score of zero

File: modelopt/torch/quantization/algorithms.py (hunk 2 above)
Comment (realAsma) [reply modelopt/torch/quantization/algorithms.py]: yes

File: modelopt/torch/quantization/algorithms.py (hunk 3 above)
Comment (realAsma) [reply modelopt/torch/quantization/algorithms.py:1159]: This is a bug. Thanks a lot for catching this. I was previously using `torch.nn.functional.kl_div` for my experiments - I missed the negative sign during my refactors.  Let me fix this.

File: modelopt/torch/quantization/algorithms.py [hunk 6]
```diff
@@ -963,8 +963,231 @@ def run_search_with_stats(self, max_weight_size, verbose=False):
...
+        )
+    return _get_softmax(logits, return_log_prob)
+
+
+def _get_kl_div_loss(
+    prob_unquant: torch.Tensor, logits_quant: torch.Tensor, lm_head: nn.Module = None
+) -> torch.Tensor:
+    log_prob_quant = _get_prob_from_logits(logits_quant, return_log_prob=True, lm_head=lm_head)
+    # We dont need to calculate the full kl div loss here, just get - p*log_q
+    return -_get_p_log_q(prob_unquant, log_prob_quant)
```
Comment (realAsma) [inline modelopt/torch/quantization/algorithms.py:1076]: @meenchen fixed the formula. Thanks a lot for catching!!

File: modelopt/torch/quantization/algorithms.py (hunk 3 above)
Comment (realAsma) [reply modelopt/torch/quantization/algorithms.py:1159]: https://github.com/NVIDIA/TensorRT-Model-Optimizer/pull/592/files#r2561243111

**Review by meenchen — APPROVED** [review]
